File: GanttChartLightLibraryDemos/Demos/Samples.Resources/WPF-CSharp/GanttChartDataGrid/Database/TaskItemsConverter.cs
using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using System.Windows.Data;
using System.Globalization;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using DlhSoft.Windows.Controls;
using System.ComponentModel;
using System.Linq;
using System.Collections.Specialized;

namespace GanttChartDataGridDatabaseSample
{
    public class TaskItemsConverter : IValueConverter
    {
        // Retrieve a collection of GanttChartItem based on Task data context.
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
        {
            IEnumerable<Task> tasks = (value as IEnumerable<Task>).OrderBy(t => t.Index);
            ObservableCollection<GanttChartItem> items = new ObservableCollection<GanttChartItem>();
            foreach (Task task in tasks)
            {
                GanttChartItem item = new GanttChartItem { Tag = task };
                SetGanttChartItem(item);
                items.Add(item);
            }
            foreach (GanttChartItem item in items)
                SetGanttChartItemPredecessors(item, items);
            items.CollectionChanged += Items_CollectionChanged;
            return items;
        }

        // When the GanttChartItem collection changes, update data context accordingly.
        private void Items_CollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
        {
            switch (e.Action)
            {
                case NotifyCollectionChangedAction.Add:
                    IEnumerable<GanttChartItem> items = sender as IEnumerable<GanttChartItem>;
                    foreach (GanttChartItem item in e.NewItems)
                    {
                        Task task = new Task { Name = item.Content as string ?? "New Task", Start = item.Start, Finish = item.Finish, Completion = item.Start, Assignments = string.Empty, Index = GetNextTaskIndex(items) };
                        item.Tag = task;
                        SetGanttChartItem(item);
                        SetGanttChartItemPredecessors(item, items);
                        Context.Tasks.AddObject(task);
                    }
                    break;
                case NotifyCollectionChangedAction.Remove:
                    foreach (GanttChartItem item in e.OldItems)
                    {
                        Task task = item.Tag as Task;
                        Context.Tasks.DeleteObject(task);
                    }
                    break;
            }
        }

        private void SetGanttChartItem(GanttChartItem item)
        {
            Task task = item.Tag as Task;
            BindingOperations.SetBinding(item, GanttChartItem.IndentationProperty, new Binding("Indentation") { Source = task, Mode = BindingMode.TwoWay });
            BindingOperations.SetBinding(item, GanttChartItem.ContentProperty, new Binding("Name") { Source = task, Mode = BindingMode.TwoWay });
            BindingOperations.SetBinding(item, GanttChartItem.StartProperty, new Binding("Start") { Source = task, Mode = BindingMode.TwoWay });
            BindingOperations.SetBinding(item, GanttChartItem.FinishProperty, new Binding("Finish") { Source = task, Mode = BindingMode.TwoWay });
            BindingOperations.SetBinding(item, GanttChartItem.CompletedFinishProperty, new Binding("Completion") { Source = task, Mode = BindingMode.TwoWay });
            BindingOperations.SetBinding(item, GanttChartItem.IsMilestoneProperty, new Binding("IsMilestone") { Source = task, Mode = BindingMode.TwoWay });
            BindingOperations.SetBinding(item, GanttChartItem.AssignmentsContentProperty, new Binding("Assignments") { Source = task, Mode = BindingMode.TwoWay });
        }

        private void SetGanttChartItemPredecessors(GanttChartItem item, IEnumerable<GanttChartItem> items)
        {
            Task task = item.Tag as Task;
            BindingOperations.SetBinding(item, GanttChartItem.PredecessorsProperty, new Binding("Predecessors") { Source = task, Converter = PredecessorItemsConverter.GetInstance(task, items) });
        }

        private int GetNextTaskIndex(IEnumerable<GanttChartItem> items)
        {
            items = items.Where(i => i.Tag != null);
            if (!items.Any())
                return 0;
            return (from i in items
                    let t = i.Tag as Task
                    select t.Index).Max() + 1;
        }

        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
        {
            throw new NotSupportedException();
        }

        public DatabaseEntities Context { get; private set; }

        public static TaskItemsConverter GetInstance(DatabaseEntities context)
        {
            return new TaskItemsConverter { Context = context };
        }
    }
}

File: GanttChartLightLibraryDemos/Demos/Samples.Resources/WPF-VisualBasic/GanttChartDataGrid/Database/TaskItemsConverter.vb
Imports System.Net
Imports System.Windows.Ink
Imports System.Windows.Media.Animation
Imports System.Globalization
Imports System.Collections.ObjectModel
Imports DlhSoft.Windows.Controls
Imports System.ComponentModel
Imports System.Collections.Specialized

Public Class TaskItemsConverter
    Implements IValueConverter

    ' Retrieve a collection of GanttChartItem based on Task data context.
    Public Function Convert(value As Object, targetType As Type, parameter As Object, culture As CultureInfo) As Object Implements IValueConverter.Convert
        Dim tasks As IEnumerable(Of Task) = (TryCast(value, IEnumerable(Of Task))).OrderBy(Function(t) t.Index)
        Dim items As New ObservableCollection(Of GanttChartItem)()
        For Each t As Task In tasks
            Dim item As GanttChartItem = New GanttChartItem With {.Tag = t}
            SetGanttChartItem(item)
            items.Add(item)
        Next t
        For Each item As GanttChartItem In items
            SetGanttChartItemPredecessors(item, items)
        Next item
        AddHandler items.CollectionChanged, AddressOf Items_CollectionChanged
        Return items
    End Function

    ' When the GanttChartItem collection changes, update data context accordingly.
    Private Sub Items_CollectionChanged(sender As Object, e As NotifyCollectionChangedEventArgs)
        Select Case e.Action
            Case NotifyCollectionChangedAction.Add
                Dim items As IEnumerable(Of GanttChartItem) = TryCast(sender, IEnumerable(Of GanttChartItem))
                For Each item As GanttChartItem In e.NewItems
                    Dim t As Task = New Task With {.Name = If(TryCast(item.Content, String), "New Task"), .Start = item.Start, .Finish = item.Finish, .Completion = item.Start, .Assignments = String.Empty, .Index = GetNextTaskIndex(items)}
                    item.Tag = t
                    SetGanttChartItem(item)
                    SetGanttChartItemPredecessors(item, items)
                    Context.Tasks.AddObject(t)
                Next item
            Case NotifyCollectionChangedAction.Remove
                For Each item As GanttChartItem In e.OldItems
                    Dim t As Task = TryCast(item.Tag, Task)
                    Context.Tasks.DeleteObject(t)
                Next item
        End Select
    End Sub

    Private Sub SetGanttChartItem(item As GanttChartItem)
        Dim t As Task = TryCast(item.Tag, Task)
        BindingOperations.SetBinding(item, GanttChartItem.IndentationProperty, New Binding("Indentation") With {.Source = t, .Mode = BindingMode.TwoWay})
        BindingOperations.SetBinding(item, GanttChartItem.ContentProperty, New Binding("Name") With {.Source = t, .Mode = BindingMode.TwoWay})
        BindingOperations.SetBinding(item, GanttChartItem.StartProperty, New Binding("Start") With {.Source = t, .Mode = BindingMode.TwoWay})
        BindingOperations.SetBinding(item, GanttChartItem.FinishProperty, New Binding("Finish") With {.Source = t, .Mode = BindingMode.TwoWay})
        BindingOperations.SetBinding(item, GanttChartItem.CompletedFinishProperty, New Binding("Completion") With {.Source = t, .Mode = BindingMode.TwoWay})
        BindingOperations.SetBinding(item, GanttChartItem.IsMilestoneProperty, New Binding("IsMilestone") With {.Source = t, .Mode = BindingMode.TwoWay})
        BindingOperations.SetBinding(item, GanttChartItem.AssignmentsContentProperty, New Binding("Assignments") With {.Source = t, .Mode = BindingMode.TwoWay})
    End Sub

    Private Sub SetGanttChartItemPredecessors(item As GanttChartItem, items As IEnumerable(Of GanttChartItem))
        Dim t As Task = TryCast(item.Tag, Task)
        BindingOperations.SetBinding(item, GanttChartItem.PredecessorsProperty, New Binding("Predecessors") With {.Source = t, .Converter = PredecessorItemsConverter.GetInstance(t, items)})
    End Sub

    Private Function GetNextTaskIndex(items As IEnumerable(Of GanttChartItem)) As Integer
        items = items.Where(Function(i) i.Tag IsNot Nothing)
        If Not items.Any() Then
            Return 0
        End If
        Return (
            From i In items
            Let t = TryCast(i.Tag, Task)
            Select t.Index).Max() + 1
    End Function

    Public Function ConvertBack(value As Object, targetType As Type, parameter As Object, culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
        Throw New NotSupportedException()
    End Function

    Private privateContext As DatabaseEntities
    Public Property Context() As DatabaseEntities
        Get
            Return privateContext
        End Get
        Private Set(ByVal value As DatabaseEntities)
            privateContext = value
        End Set
    End Property

    Public Shared Function GetInstance(context As DatabaseEntities) As TaskItemsConverter
        Return New TaskItemsConverter With {.Context = context}
    End Function
End Class
